### PR TITLE
Make tests basis-independent

### PR DESCRIPTION
**Pull request by pbrubeck**
---
name: "Make tests basis-independent"
description: ''
title: ''
labels: ''
assignees: ''

---

# Description
We need to drop the first commit and re-run CI with the equispaced variant.

A few observations: 

1. `VertexLimiter` requires functions in `Broken(mesh.coordinates.function_space()) `
2. Some tests like the one for `EquationBC` use a periodic sine wave manufactured solution, this could show the asymptotic convergence rate sooner with equispaced points (these points have the best approximation properties for periodic functions).
3. The tests for the adjoint of cross-mesh interpolation were testing for inequality of the output against the exact target `Cofunction`, and the source `Function`. The issue with this is that the absolute tolerance used in the dual comparison was the same used for primal comparsion, and the change to the spectral basis was changing the magnitude of the coefficients of the `Cofunctions` below the absolute tolerance. These inequality tests are quite meaningless and should be replaced with the same approach used in the multigrid `restrict` [tests](https://github.com/firedrakeproject/firedrake/blob/a07f37306fa7816b868bd19bd9839e67eb2bf968/tests/multigrid/test_grid_transfer.py#L135). We now test that `<Ix, y> == <x, adjoint(I)y>` for a choice of `x \in V_src` and `y \in V_dest*`.

## Associated Pull Requests:
- Make a list (with links!)
- [Like this PR#691](https://github.com/OP2/PyOP2/pull/691)

## Fixes Issues:
- Make a list of issues (with links!)
- [Like this Issue#2824](https://github.com/firedrakeproject/firedrake/issues/2824)

If issues are fixed by this PR, prepend each of them with the word
"fixes", so they are automatically closed when this PR is merged. For
example "fixes #123, fixes #456".

# Checklist for author:

- [ ] I have linted the codebase (`make lint` in the `firedrake` source directory).
- [ ] My changes generate no new warnings.
- [ ] All of my functions and classes have appropriate docstrings.
- [ ] I have commented my code where its purpose may be unclear.
- [ ] I have included and updated any relevant documentation.
- [ ] Documentation builds locally (`make linkcheck; make html; make latexpdf` in `firedrake/docs` directory)
- [ ] I have added tests specific to the issues fixed in this PR.
- [ ] I have added tests that exercise the new functionality I have introduced
- [ ] Tests pass locally (`pytest tests` in the `firedrake` source directory) (useful, but not essential if you don't have suitable hardware).
- [ ] I have performed a self-review of my own code using the below guidelines.

# Checklist for reviewer:

- [ ] Docstrings present
- [ ] New tests present
- [ ] Code correctly commented
- [ ] No bad "code smells"
- [ ] No issues in parallel
- [ ] No CI issues (excessive parallelism/memory usage/time/warnings generated)
- [ ] Upstream/dependent branches and PRs are ready

Feel free to add reviewers if you know there is someone who is already aware of this work.

Please open this PR initially as a draft and mark as ready for review once CI tests are passing.

### CODE REVIEW

**Comment by pbrubeck** [comment]
A few observations: 

1. `VertexLimiter` requires functions in `Broken(mesh.coordinates.function_space()) `
2. Some tests like the one for `EquationBC` use a periodic sine wave manufactured solution, this could show the asymptotic convergence rate sooner with equispaced points (these points have the best approximation properties for periodic functions).
3. I am not sure if we ever test that the 1D fiat dual basis is sorted in ascending order, as we rely on this to simplfy tabulations onto these points https://github.com/FInAT/FInAT/blob/master/finat/spectral.py#L30. The FIAT PR https://github.com/firedrakeproject/fiat/pull/46 will include such test, along with tests for symmetry.

File: firedrake/interpolation.py [hunk 1]
```diff
@@ -617,6 +617,9 @@ def interpolate(
             self.point_eval_interpolator.interpolate(
                 f_src_at_src_node_coords, transpose=True, output=output
             )
+            # FIXME we cannot yet use actual cofunctions, see the note above
+            if not isinstance(output, firedrake.Cofunction):
+                output = output.riesz_representation("l2")
```
Comment (dham) [inline firedrake/interpolation.py:622]: This looks like a bug.

File: tests/regression/test_interpolate.py [hunk 2]
```diff
@@ -231,7 +231,7 @@ def test_cellvolume():
 
 def test_cellvolume_higher_order_coords():
     m = UnitTriangleMesh()
-    V = VectorFunctionSpace(m, 'P', 3)
+    V = VectorFunctionSpace(m, "CG", 3)
```
Comment (dham) [inline tests/regression/test_interpolate.py:234]: Please don't

**Review by dham — CHANGES_REQUESTED** [review]
Just one trivial change.

File: firedrake/interpolation.py (hunk 1 above)
Comment (nbouziani) [reply firedrake/interpolation.py]:
```
            f_src_at_dest_node_coords_dest_mesh_decomp = firedrake.Cofunction(
                self.to_input_ordering_interpolator.V.dual()
            )
```

Yes, I spotted this in #2297.

Comment (pbrubeck) [reply firedrake/interpolation.py]: It should now be dual safe. @nbouziani does this look good?

**Pull request merged**